### DISCUSSION TITLE
Q: Performance/size impact of custom key comparisons

### DISCUSSION THREAD

**Discussion by snej** [erthink/libmdbx#356] [Q&A] [answered]
Hope you don't mind another question. Does using a custom `MDBX_cmp_func` key comparison function lower performance (aside from the time taken by the custom function itself) or increase database size?

It looks like there's no overhead for calling the function — both built-in and custom comparisons are called through the `md_cmp` function pointer — unless I missed something.

But from the bit I know about B-trees, I suspect that key prefix/suffix compression might not be possible if the keys aren't in a predefined order known to the B-tree manager. So if my custom-sorted keys have a lot of common prefixes, they might not be packed into nodes as efficiently. True?

**Comment by erthink** [comment, marked as answer]
> Does using a custom `MDBX_cmp_func` key comparison function lower performance (aside from the time taken by the custom function itself) or increase database size?

Neither both.

> It looks like there's no overhead for calling the function — both built-in and custom comparisons are called through the `md_cmp` function pointer — unless I missed something.

Yes, that's right, and you didn't miss anything.

> But from the bit I know about B-trees, I suspect that key prefix/suffix compression might not be possible if the keys aren't in a predefined order known to the B-tree manager.

Yes, but neither MDBX nor LMDB uses key compression. Key compression saves space for a large number of records, but results in both a loss of performance and extreme code complexity. The best solution is not to use long keys.

> So if my custom-sorted keys have a lot of common prefixes, they might not be packed into nodes as efficiently. True?

No, since no key compression here. However, **I strongly recommend not using custom comparisons** since in case MDBX you will be unable use mdbx_dump/mdbx_restore tools and will partially loose mdbx_chk functionality.

My recommendation is to convert your custom key to a linear sequence of bytes so that you can use one of the built-in comparison options. For instance, you can reorder/swap key's bytes before put record into DB, and then reorder/swap ones back after reading for recovery an initial custom key.

At the moment, the ability to use custom comparisons is not removed, only because there are several applications that have been using it for a long time and can not refuse, since this requires conversion of end-user databases and more earlier backups (which read-only mostly).

**Comment by snej** [comment]
>The best solution is not to use long keys.

I'm using DBIs to implement indexes of structured data from a separate main DBI.  So the length of the keys depends on the length of the property values that clients decide to index.

>I strongly recommend not using custom comparisons

It's unavoidable in some cases. For example, collating Unicode strings uses some very complex algorithms (fortunately implemented by ICU!) and can't be expressed by transforming the data. I will definitely need to support this.

Indexing arbitrary numbers (mixtures of int64s and floats) is also tricky. You can't store them in the index all as integers or you lose FP precision, and you can't store them all as doubles or you lose precision of very large int64s. It's also desirable to not use 8 bytes for every number, because in the real world most of them are much smaller than that. There are encodings that can do this (similar to varints) but they're a pain to deal with (I've done it in the past.)

**Comment by erthink** [comment]
> It's unavoidable in some cases. For example, collating Unicode strings uses some very complex algorithms (fortunately implemented by ICU!) and can't be expressed by transforming the data. I will definitely need to support this.

For Unicode you need 'collation for machine' aka 'Generate ICU Sortkeys', which is a basic ICU feature. This is a regular way for DBMS (i.e. Oracle, MSSQL, etc). One of benefits is sortkeys are shorter than unicode strings and more handy for machine (i.e. 32-bit aligned). On the other hand, actual characteristics is highly dependent from implementation, and ICU is not a winner here (AFAIK).

> Indexing arbitrary numbers (mixtures of int64s and floats) is also tricky. You can't store them in the index all as integers or you lose FP precision, and you can't store them all as doubles or you lose precision of very large int64s. It's also desirable to not use 8 bytes for every number, because in the real world most of them are much smaller than that. There are encodings that can do this (similar to varints) but they're a pain to deal with (I've done it in the past.)

Hm, I can't agree with you. I don't think it's a good idea to mix different data types (especially without limiting their ranges) in the same index. However, it is enough to follow some of the ideas illustrated in msgpack and Tarantool (uses comparison without unpacking).

If you don't mix different types in the same index, there are no problems. Moreover, encoding like msgpack will help reduce the index size slightly.
Otherwise, anyway, you will have to limit the range of values (store numbers like lua and/or javascript), or add a prefix to distinguish ones.

**Comment by snej** [comment]
I'll look into ICU Sortkeys ... thanks!

>I don't think it's a good idea to mix different data types (especially without limiting their ranges) in the same index.

I have no control over that, unfortunately! The data source is JSON and provided by the end user/developer. (And technically JSON only has a single number type, but we parse them to either int64 or double to preserve fidelity, depending on whether there's a decimal point or "e".)

>However, it is enough to follow some of the ideas illustrated in msgpack and Tarantool

I've partially implemented this in the past. It's simple for whole numbers, surprisingly tricky for integers, and I didn't get around to figuring out how to handle real numbers. Just now I took a look at msgpack and Tarantool, but didn't find any docs of sortable number encodings.

Anyway. It's considerably more work to make memcmp-compatible keys, and it sounds like the only drawback of custom comparisons is the inability to use some tools, so I'd prefer to stick with those. But I understand why you disagree :)

**Comment by erthink** [comment]
> The data source is JSON and provided by the end user/developer. (And technically JSON only has a single number type, but we parse them to either int64 or double to preserve fidelity, depending on whether there's a decimal point or "e".)

Seems you should reconsider about JSON numbers and a little dive into the details of JavaScript implementation (especially to end of [page 6 of RFC-7159](https://tools.ietf.org/html/rfc7159#page-6), `Number.MIN_SAFE_INTEGER` and `Number.MAX_SAFE_INTEGER`). In general, 8 bytes is enough to store any JSON number conformed RFC-7159. Moreover it is easy to make common cross-comparable form for both types (i.e. float and integer).

Pseudocode:
```
double_to_key(v: double) {
  bool sign = ieee754_get_sign(v);
  int11 exponenta = ieee754_get_exp(v);
  int52 mantissa = ieee754_get_mantissa(v);
  return encode(sign, exponenta, mantissa);
}

int64_to_key(v: int64) {
  if (v < -Number.MIN_SAFE_INTEGER || v > Number.MAX_SAFE_INTEGER)
    return double_to_key(value);
  bool sign = (v < 0) ? true : false;
  int11 exponenta = 0;
  int52 mantissa = (v < 0) ? -v : v;
  return encode(sign, exponenta, mantissa);
}
```

**Comment by erthink** [comment]
@snej, can we close this issue or there still are other questions?

**Comment by snej** [comment]
The RFC "_allows_ implementations to set limits on the range and precision of numbers accepted." Our implementation allows 64-bit integers. I know JavaScript treats all numbers as doubles, but this isn't JavaScript.

Your pseudocode doesn't work when comparing negative numbers; e.g. it will say -3 > -2. It also doesn't compare doubles with ints correctly, since a double greater than two has a positive exponent while an integer always has zero. These are fixable problems of course.

No more questions on this topic; but count me as strongly opposed to the removal of custom comparison functions.

**Comment by erthink** [comment]
@snej, I think I accidentally confused you in the details, without showing the main thing.
The following is a simplified C++ example of how you can get unified keys/indexes (both for double and int64 with a limited range within 53 bits) for built-in comparators (`MDBX_INTEGERKEY` or `MDBX_INTEGERDUP`). 

```
enum {
  IEEE754_DOUBLE_SIGN_MASK = UINT64_C(0x8000000000000000),
  JSON_MIN_SAFE_INTEGER = -((INT64_C(1) << 53) - 1),
  JSON_MAX_SAFE_INTEGER = (INT64_C(1) << 53) - 1
};

static uint64_t double_to_key(const double d) {
  uint64_t i;
  assert(sizeof(d) == sizeof(i));
  memcpy(&u, &d, sizeof(i));
  if (i < 0) {
    // negative
    return (uint64_t)i;
  }
  // positive
  return IEEE754_DOUBLE_SIGN_MASK + i;
}

static uint64_t int53_to_key(const int64_t i) {
  assert(i >= JSON_MIN_SAFE_INTEGER && i <= JSON_MAX_SAFE_INTEGER);
  // for the sake of simplicity: 
  // conversion to double to avoid need of an encode() function.
  return double_to_key(double(i));
}
```
**UPDATE: Minor fix for negative-case (adds +1 to result).**
